lib/filter/QubitTransactionFilter.class.php

```diff
@@ -32,6 +34,45 @@ public static function getConnection()
     return self::$connection;
   }
 
+  protected function retry()
+  {
+    // If we've hit the retry limit, abort and return false
+    if (self::$retry++ > self::$retryLimit)
+    {
+      return false;
+    }
+
+    // Log a warning
+    if (sfConfig::get('sf_logging_enabled'))
+    {
+      $this->context->getLogger()->warning(
+        sprintf('Encountered a SQL transaction deadlock, retry %d of %d',
+          self::$retry, self::$retryLimit
+        )
+      );
+    }
+
+    // Get the current action instance
+    $actionInstance = $this->context
+      ->getController()
+      ->getActionStack()
+      ->getLastEntry()
+      ->getActionInstance();
+
+    // Create a new filter chain and reload config
+    $filterChain = new sfFilterChain();
+    $filterChain->loadConfiguration($actionInstance);
+
+    $this->context->getEventDispatcher()->notify(
+      new sfEvent($this, 'transaction.retry')
+    );
+
+    // Execute whole filter chain again
+    $filterChain->execute();
```

Comment:
:mage:

That's a nice one @djjuhasz! I didn't know you could something like this and I have a few questions ...

- It's rare to see actions chained together and I wonder if it even matters but, should we get the first action entry to fully retry the request?
- What would happen when the request cycle involves other processes apart from the DB transaction? I'm thinking in DO uploads for example, would the file be copied again, etc.?

Reply:
Another minor thing, why the event dispatcher? I'd not add it unless it's listened somewhere else.

Reply:
>     * It's rare to see actions chained together and I wonder if it even matters but, should we get the first action entry to fully retry the request?

My current theory is that the whole filter chain is executed for *every* action, and that only the current (last) action is relevant to $filter->execute(). This theory is based on:

1. The `$filterChain->loadConfiguration($actionInstance)` requires a single$actionInstance,
2. The `sfController::forward()` method adds a new action to the action stack AND "creates and executes a new $filterChain":https://github.com/artefactual/atom/blob/1c70ba36c0ad85ed0fd5cef8a64508ae41c207d0/vendor/symfony/lib/controller/sfController.class.php#L224,
3. The `sfExecutionFilter()` only executes the "last action on the stack":https://github.com/artefactual/atom/blob/qa/2.6.x/vendor/symfony/lib/filter/sfExecutionFilter.class.php#L35 (it's where I got the code to get the $actionInstance in QubitTransactionFilter)

Reply:
>What would happen when the request cycle involves other processes apart from the DB transaction? I'm thinking in DO uploads for example, would the file be copied again, etc.?

The request is processed before the filter chain is executed, but all of the code in the filter chain and action execution stages will be re-run.  I think it's necessary to run the filter/action code again when redoing the transaction to avoid problems like the nested set values being stale, and because there is no way (that I know of) with symfony or Propel to *just* replay all of the database read/writes in isolation from the execution code.

Reply:
> Another minor thing, why the event dispatcher? I'd not add it unless it's listened somewhere else.

Yes, fair enough.  I was copying the pattern from https://github.com/artefactual/atom/blob/qa/2.6.x/vendor/symfony/lib/controller/sfController.class.php#L224-L238 but in the absence of listeners, the event dispatcher notification is not doing anything.

Reply:
> By re-executing the filter chain, I believe steps 7-10 are re-run, then the response is output (step 11).

Yeah, what worries me the most is the re-execution of step 9 but lets hope the actions that may cause a DB deadlock don't do a lot of not idempotent work outside the DB changes.

Reply:
Yes, there's a lot of magic going on behind the curtain here.  I wish I knew a practical way to determine if this approach is creating regressions, but without good test coverage I can't see how. 🤷 As you say, hopefully the code execution is effectively idempotent outside of the DB transaction boundaries.

Reply:
I guess worst-case scenario it won't take a lot of work to undo the transaction retry code, so if testing does turn up regressive behaviour, reversing the changes can be the exit strategy.